### ISSUE TITLE
0.9.4: build fails

### ISSUE THREAD

**Issue by kloczek** [iovisor/bpftrace#1126]
```
[ 14%] Building CXX object src/CMakeFiles/bpftrace.dir/btf.cpp.o
cd /home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src && /usr/bin/g++  -DBPFTRACE_VERSION=\"v0.9.4\" -DHAVE_BCC_CREATE_MAP -DHAVE_BCC_PROG_LOAD -DHAVE_BFD_DISASM -DHAVE_NAME_TO_HANDLE_AT=1 -DLIBBFD_DISASM_FOUR_ARGS_SIGNATURE -I/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src -I/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/ast -I/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4 -I/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/resources/../src -isystem /usr/include/bcc  -O2 -g -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fstack-protector-strong -grecord-gcc-switches -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1 -m64 -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection  -Os -O2 -g -DNDEBUG   -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Wall -Wextra -Wundef -Wpointer-arith -Wcast-align -Wwrite-strings -Wcast-qual -Wswitch-default -Wunreachable-code -Wdisabled-optimization -std=c++14 -o CMakeFiles/bpftrace.dir/btf.cpp.o -c /home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.cpp
In file included from /home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.cpp:1:
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.h:25:8: error: ‘string’ in namespace ‘std’ does not name a type
   25 |   std::string c_def(std::unordered_set<std::string>& set);
      |        ^~~~~~
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.h:6:1: note: ‘std::string’ is defined in header ‘<string>’; did you forget to ‘#include <string>’?
    5 | #include <unordered_set>
  +++ |+#include <string>
    6 |
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.h:26:8: error: ‘string’ in namespace ‘std’ does not name a type
   26 |   std::string type_of(const std::string& name, const std::string& field);
      |        ^~~~~~
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.h:26:3: note: ‘std::string’ is defined in header ‘<string>’; did you forget to ‘#include <string>’?
   26 |   std::string type_of(const std::string& name, const std::string& field);
      |   ^~~
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.h:27:8: error: ‘string’ in namespace ‘std’ does not name a type
   27 |   std::string type_of(const btf_type* type, const std::string& field);
      |        ^~~~~~
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.h:27:3: note: ‘std::string’ is defined in header ‘<string>’; did you forget to ‘#include <string>’?
   27 |   std::string type_of(const btf_type* type, const std::string& field);
      |   ^~~
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.cpp:312:13: error: no declaration matches ‘std::string bpftrace::BTF::c_def(std::unordered_set<std::__cxx11::basic_string<char> >&)’
  312 | std::string BTF::c_def(std::unordered_set<std::string>& set __attribute__((__unused__))) { return std::string(""); }
      |             ^~~
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.cpp:312:13: note: no functions named ‘std::string bpftrace::BTF::c_def(std::unordered_set<std::__cxx11::basic_string<char> >&)’
In file included from /home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.cpp:1:
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.h:12:7: note: ‘class bpftrace::BTF’ defined here
   12 | class BTF
      |       ^~~
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.cpp:314:13: error: no declaration matches ‘std::string bpftrace::BTF::type_of(const string&, const string&)’
  314 | std::string BTF::type_of(const std::string& name __attribute__((__unused__)),
      |             ^~~
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.cpp:314:13: note: no functions named ‘std::string bpftrace::BTF::type_of(const string&, const string&)’
In file included from /home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.cpp:1:
/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4/src/btf.h:12:7: note: ‘class bpftrace::BTF’ defined here
   12 | class BTF
      |       ^~~
make[2]: *** [src/CMakeFiles/bpftrace.dir/build.make:105: src/CMakeFiles/bpftrace.dir/btf.cpp.o] Error 1
make[2]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4'
make[1]: *** [CMakeFiles/Makefile2:1203: src/CMakeFiles/bpftrace.dir/all] Error 2
make[1]: Leaving directory '/home/tkloczko/rpmbuild/BUILD/bpftrace-0.9.4'
```

**Comment by fbs** [comment]
What distro/gcc/llvm/clang version are you using here?

**Comment by kloczek** [comment]
gcc 10.0.1.

**Comment by caringi** [comment]
I faced the same problem while trying to build it on Fedora Rawhide (32), with gcc 10.0.1.

The following patch fixes the build:

```
diff --git a/src/btf.h b/src/btf.h
index 05b56ff..8856bfc 100644
--- a/src/btf.h
+++ b/src/btf.h
@@ -1,6 +1,7 @@
 #pragma once
 
 #include <linux/types.h>
+#include <string>
 #include <unistd.h>
 #include <unordered_set>
```

I'm going to submit a PR...

**Comment by kloczek** [comment]
Confirmed.

With that patch everything builds corectly.

**Comment by fbs** [comment]
Does this also happen on fedora 31 (does that use btf at all)?

**Comment by kloczek** [comment]
My system is half fedora rawhide and half my own.

I've not been using yet just compiled binaries :)

**Comment by caringi** [comment]
The explanation of why this problem appeared with gcc 10 is here:

> **Porting to GCC 10**
> 
> The GCC 10 release series differs from previous GCC releases in a number of ways. Some of these are a result of bug fixing, and some old behaviors have been intentionally changed to support new standards, or relaxed in standards-conforming ways to facilitate compilation or run-time performance.
> 
> Some of these changes are user visible and can cause grief when porting to GCC 10. This document is an effort to identify common issues and provide solutions. Let us know if you have suggestions for improvements! 
> 
> **Header dependency changes**
>
> Some C++ Standard Library headers have been changed to no longer include the <stdexcept> header. As such, C++ programs that used components defined in <stdexcept> or <string> without explicitly including the right headers will no longer compile.
> 
> Previously components such as std::runtime_error, std::string and std::allocator were implicitly defined after including unrelated headers such as <array> and <optional>. Correct code should include the appropriate headers for the classes being used. 

https://gcc.gnu.org/gcc-10/porting_to.html

**Comment by fbs** [comment]
Closing this as  #1127   fixed this and we have  #1130 fix gcc warnings
